src/heap_utils.cc

```diff
@@ -231,19 +234,178 @@ class BufferOutputStream : public v8::OutputStream {
   std::unique_ptr<JSString> buffer_;
 };
 
+namespace {
+class FileOutputStream : public v8::OutputStream {
+ public:
+  explicit FileOutputStream(FILE* stream) : stream_(stream) {}
+
+  virtual int GetChunkSize() {
+    return 65536;  // big chunks == faster
+  }
+
+  virtual void EndOfStream() {}
+
+  virtual WriteResult WriteAsciiChunk(char* data, int size) {
+    const size_t len = static_cast<size_t>(size);
+    size_t off = 0;
+
+    while (off < len && !feof(stream_) && !ferror(stream_))
+      off += fwrite(data + off, 1, len - off, stream_);
+
+    return off == len ? kContinue : kAbort;
+  }
+
+ private:
+  FILE* stream_;
+};
+
+class StreamBaseOutputStream : public AsyncWrap,
+                               public StreamBase,
+                               public v8::OutputStream {
+ public:
+  StreamBaseOutputStream(
+      Environment* env,
+      const HeapSnapshot* snapshot,
+      v8::Local<v8::Object> obj) :
+      AsyncWrap(env, obj, AsyncWrap::PROVIDER_HEAPSNAPSHOT),
+      StreamBase(env),
+      snapshot_(snapshot),
+      ended_(false) {
+    MakeWeak();
+    StreamBase::AttachToObject(GetObject());
+  }
+
+  ~StreamBaseOutputStream() override {
+    const_cast<HeapSnapshot*>(snapshot_)->Delete();
+  }
+
+  int GetChunkSize() override {
+    return 65536;  // big chunks == faster
+  }
+
+  void EndOfStream() override {
+    ended_ = true;
```

Comment:
One possibly worthwhile optimization w.r.t. memory consumption (since snapshots can be big) is to opportunistically delete the snapshot in this method instead of always deferring to the destructor.

Extra nice is that you can get rid of `ended_` and just check `snapshot_ != nullptr` in IsAlive().